Use AssemblyLoadContext-aware caches in TypeDescriptor to support unloading of assemblies cached by TypeDescriptor

alexey-zakharov opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/30656 - TypeDescriptor caches keeping strong references to the types from collectible assemblies.

**Motivation**

`TypeDescriptor` uses static caches to amortize attributes retrieval and store user providers. The caches use `Type`-to-object approach and capture strong reference to the Type and dependent objects. That also includes `ReflectTypeDescriptionProvider`) implementation. Strong reference to the Type prevents collectible types and thus collectible assemblies from being collected and unloaded.

### Changes

The PR applies the "context aware table" pattern used in coreclr codebase to split caching of collectible and non-collectible types into 2 tables - standard Hashtable/Dictionary for non-collectible types and WeakHashtable/ConditionalWeaktable for collectible types.

- Introduced `ContextAwareConcurrentHashtable` and `ContextAwareHashtable` helper classes which leverage `DependentHandle`/`WeakReference` for collectible types.
- Used context aware tables `TypeDescription` and `ReflectTypeDescriptionProvider` implementations to ensure there are no strong references from static caches to types from unloadable assemblies.

Note that the collectible Type used as a key will be only collected once `AssemblyLoadContext.Unload` is called since the [type instance](https://github.com/dotnet/runtime/blob/27604b57bd2e9c9aa46d005db7c4e387d461b5b6/src/coreclr/vm/methodtable.h#L363) for collectible types is allocated by `LoaderAllocator` and stored in its `m_slots` table for the whole lifetime of `LoaderAllocator`. Thus all custom collectible providers will be kept alive until the point when their parent AssemblyLoadContext is explicitly unloaded.

The performance impact of the change may be observed for the collectible types - dereferencing `DependentHandle` and `WeakReference` adds overhead and may introduce additional entries in the handle tables. And the cost of calling `MemberInfo.IsCollectible` is added to all queries.

The PR addresses ~~https://github.com/dotnet/runtime/issues/114620~~ https://github.com/dotnet/runtime/issues/30656 issue.


**Testing**

- Added `TypeDescriptorTests.TypeDescriptor_DoesNotKeepUnloadableTypes` test to validate caching for collectible types.
- Validated System.ComponentModel.TypeConverter.Tests locally.
- Validated no ALC leaks are present using a custom reporting tool.

alexey-zakharov committed:
﻿Use separate hashtables for collectible types to support assembly unloadability for the TypeDescriptor class

jkotas commented:
> caches are cleared on AssemblyLoadContext unloading.

MetadataUpdateHandler is not meant to be used to prepare caches for AssemblyLoadContext unloading.

jkotas commented:
> the potential alternative could be to use WeakHashtable instead. However this would come with performance costs

The simplest way to avoid the overhead is to skip the caching completely for collectible types. We do that in number of places in this repo.

alexey-zakharov commented:
> The simplest way to avoid the overhead is to skip the caching completely for collectible types. We do that in number of places in this repo.

thanks for the suggestion!
we could skip caching completely, but I would expect potential json serialization performance regression for types from collectible assemblies
I'll update the PR to do this instead

alexey-zakharov commented:
> If the performance regression is not acceptable, it is an option to have two caches - one for collectible types and one for non-collectible types. It avoids the performance overhead of weak handles for non-colllectible types.

@jkotas to clarify - would you be ok with a solution that keeps existing cache for types in non-collectible ALCs and WeakHashtable for the types in collectible ALCs (for both TypeDescriptor and ReflectTypeDescriptionProvider cache levels), correct?

I like this approach, as it gives correctness and performance at a relatively low cost of calling Type.IsCollectible 😄

steveharter commented on src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs:
```diff
@@ -2530,6 +2530,32 @@ public static void Refresh(Assembly assembly)
             }
         }
 
+        internal static void ClearCache(Type type)
```
This and the other ClearCache() below will need to lock on:
```cs
lock (s_commonSyncObject)
```

Adding items to `s_defaultProviderInitialized` is already protected this way and so should removing.

alexey-zakharov replied:
thanks @steveharter !
to clarify - are you ok with the approach in the PR or should I split caches into 2 - one for non-collectible types and one for collectible types (then we don't need ClearCache changes)

steveharter commented:
I'm not sure on what issue this PR is current solving. Currently it depends on hot reload APIs to be called to clear caches, but that already works for hot reload purposes AFAIK. It doesn't solve the more general issue of unloading a collectible ALC.

jkotas commented:
> would you be ok with a solution that keeps existing cache for types in non-collectible ALCs and WeakHashtable for the types in collectible ALCs

It would be fine with me. We have similar split in other places.

alexey-zakharov commented:
> I'm not sure on what issue this PR is current solving. Currently it depends on hot reload APIs to be called to clear caches, but that already works for hot reload purposes AFAIK. It doesn't solve the more general issue of unloading a collectible ALC.

@steveharter  yes, I'm going to rework it to split the caches next week.
the reason we went with reusing HotReload api is that the places that we've found leaks so far were matching the [MetadataUpdateHandler](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.metadata.metadataupdatehandlerattribute?view=net-9.0) usages.
I think moving to weak table would remove the need of relying on the MetadataUpdateHandler.

Although there is one case which may need a cache clear approach - `ArrayPool<>.Shared` holds references to the unloading ALCs in the returned pools. And while it is purged eventually, the purge speed it not sufficient to not run out of memory due to accumulating ALCs. But I'll raise this issue separately - there are multiple possible solutions to consider

alexey-zakharov commented:
@steveharter @jkotas I've updated the implementation to use `WeakHashtable` and `ContextAware*` patterns to have a separate caches for collectible types.

That give us a good balance between functional correctness and performance as:

- `WeakReference` ensures there are no gaps between AssemblyLoadContext unload and repopulating the cache.
- The best case performance overhead (non-unloadable assemblies) is effectively `MemberInfo.IsCollectible` call only.

alexey-zakharov commented on src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReflectTypeDescriptionProvider.cs:
```diff
@@ -49,10 +49,12 @@ internal sealed partial class ReflectTypeDescriptionProvider : TypeDescriptionPr
-        private static Hashtable? s_extendedPropertyCache;
+        // The cache uses ContextAwareHashtable to ensure collectible types are stored via WeakReference
+        // to allow for collectible AssemblyLoadContexts to be unloaded.
+        private static ContextAwareHashtable<PropertyDescriptor[]>? s_propertyCache;
```
if we want to keep a non-generic version to limit the amount of jit we need to do, I'm happy to make `ContextAwareHashtable` non-generic

alexey-zakharov commented on src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs:
```diff
@@ -50,19 +48,19 @@ public sealed class TypeDescriptor
 
         // A direct mapping from type to provider.
-        private static readonly ConcurrentDictionary<Type, TypeDescriptionNode> s_providerTypeTable = new ConcurrentDictionary<Type, TypeDescriptionNode>();
+        private static readonly ContextAwareConcurrentDictionary<Type, TypeDescriptionNode> s_providerTypeTable = new ContextAwareConcurrentDictionary<Type, TypeDescriptionNode>(useWeakValue : true);
```
value is stored as a `WeakReference` for the `s_providerTypeTable` and is repopulated from the `s_providerTable` if collected, so functionally there is no danger of loosing custom providers added with `AddProvider` method, but from performance perspective, perhaps switching to `DependentHandle` would be better, cc @ericstj

alexey-zakharov commented on src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs:
```diff
@@ -339,6 +337,7 @@ private static void CheckDefaultProvider(Type type)
                 return;
             }
 
+            // Take the lock to check for the collectible type case and add default provider if nothing is yet registered.
```
```suggestion
```

ericstj reviewed (commented):
This looks clean.  Just a few suggestions and questions.

alexey-zakharov commented on src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ContextAwareConcurrentHashtable.cs:
```diff
@@ -0,0 +1,137 @@
+    /// Uses ConditionalWeakTable for the collectible keys (if MemberInfo.IsCollectible is true) and
+    /// ConcurrentDictionary for non-collectible keys.
+    /// </summary>
+    internal sealed class ContextAwareConcurrentHashtable<TKey, [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor)] TValue> : IEnumerable<KeyValuePair<TKey, TValue>>
```
good suggestion!

alexey-zakharov commented on src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ContextAwareConcurrentHashtable.cs:
```diff
@@ -0,0 +1,137 @@
+
+            set
+            {
+                if (!key.IsCollectible)
```
this is mostly to highlight the "default" most common case first - it makes code smaller in some cases (e.g. [TryGetValue](https://github.com/dotnet/runtime/pull/114619/files#diff-10ac6ede51b832797d2b9a98abac3b66ce033d384533b9c7e612837871cc4073R50))

alexey-zakharov commented on src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ContextAwareConcurrentHashtable.cs:
```diff
@@ -0,0 +1,137 @@
+            {
+                _defaultEnumerator.Reset();
+                _collectibleEnumerator.Reset();
+                Current = default;
```
well spotted!

alexey-zakharov commented on src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/WeakHashtable.cs:
```diff
@@ -23,15 +23,37 @@ internal WeakHashtable() : base(s_comparer)
+        /// Override of GetEnumerator that returns enumerator which skips not alive keys
+        /// during enumeration.
+        /// </summary>
+        public override IDictionaryEnumerator GetEnumerator() => new WeakHashtableEnumerator(base.GetEnumerator());
```
do you mean for safety reasons? I could do it, but I was juggling between - safety and "this class was only used in few specific places, keep size overhead small" 😄

alexey-zakharov replied:
define IEnumerable.GetEnumerator()

alexey-zakharov commented on src/libraries/System.ComponentModel.TypeConverter/tests/UnloadableTestTypes/UnloadableTestTypes.cs:
```diff
@@ -0,0 +1,22 @@
+namespace UnloadableTestTypes
+{
+    [SimpleType]
+    public class SimpleType
```
awesome idea!

alexey-zakharov committed:
Addressed codereview feedback

alexey-zakharov committed:
Added a test for custom provider and updated WeakHashTable to use ConditionalWeakTable to ensure values can be also colelcted

alexey-zakharov commented:
> @alexey-zakharov have you had a chance to look?

not yet, thanks for the ping! 
we've been using the changes internally and didn't spot any issues, however the custom provider case that you've mentioned required a change since WeakHashtable should also use a weak value

alexey-zakharov commented on src/libraries/System.ComponentModel.TypeConverter/System.ComponentModel.TypeConverter.slnx:
```diff
@@ -263,6 +263,9 @@
       <Build Solution="Checked|x86" Project="false" />
     </Project>
+    <Project Path="tests/UnloadableTestTypes/UnloadableTestTypes.csproj">
+      <BuildType Solution="Checked|*" Project="Release" />
```
aligned with other project references

alexey-zakharov commented on src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/WeakHashtable.cs:
```diff
@@ -168,5 +190,65 @@ public override bool Equals(object? o)
+                get
+                {
+                    object key = _baseEnumerator.Key;
+                    return ((EqualityWeakReference)key).Target!;
```
I've trimmed the api to IEnumerable and only used hashtable methods

alexey-zakharov commented on src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ContextAwareConcurrentHashtable.cs:
```diff
@@ -0,0 +1,137 @@
+                }
+                else
+                {
+                    _collectibleTable.AddOrUpdate(key, value!);
```
applied the suggestion manually

alexey-zakharov commented on src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ContextAwareHashtable.cs:
```diff
@@ -0,0 +1,40 @@
+                }
+                else
+                {
+                    _collectibleTable.AddOrUpdate(key, value!);
```
applied the suggestion manually

ericstj reviewed (approved):
Thank you for addressing the feedback.  This is looking good, just one minor renaming suggestion for consistency.

alexey-zakharov committed:
Renamed ContextAwareHashtable to CollectibleKeyHashtable

alexey-zakharov commented on src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ContextAwareHashtable.cs:
```diff
@@ -0,0 +1,43 @@
+    /// This ensures that collectible <see cref="MemberInfo"/> instances (such as those from collectible assemblies) do not prevent their assemblies from being unloaded.
+    /// </para>
+    /// </summary>
+    internal sealed class ContextAwareHashtable
```
yes, sorry, missed the hint :)

alexey-zakharov commented:
> Thank you for addressing the feedback. This is looking good, just one minor renaming suggestion for consistency.

awesome, thanks for the help with the PR!

alexey-zakharov commented:
@ericstj do you have any more suggestions? should the PR be merged? 😄

alexey-zakharov committed:
Merge branch 'main' into dotnet-upstream/fix-typedescription-alc-leak

alexey-zakharov committed:
Merge branch 'main' into dotnet-upstream/fix-typedescription-alc-leak

alexey-zakharov committed:
Merge branch 'main' into dotnet-upstream/fix-typedescription-alc-leak

ericstj reviewed (approved):
Sorry for taking a while in getting back to you, I missed the notification on this.

ericstj commented:
@alexey-zakharov do you need this in net10.0 or is main  sufficient?

ericstj commented:
I would definitely like to see some bake time for this change before it makes it to 10.0 - the challenge is that the bar for 10.0 will only be getting higher from here on.  @vitek-karas do you have opinions on this?  How can we gain confidence in this fix in the scenarios that matter for https://github.com/dotnet/runtime/issues/30656?

vitek-karas commented:
I agree that this change is not the riskier side - mostly because it's used under the hood by many components and potential issues might take a while to show up.
